Issue/3016 - Settings for jumpTo

WeirdRubberDuck opened the pull request:
Add a lua script for jumpTo, and a prpoerty that controls how long the fading should take. 

Closes #3016 together with a PR in the webgui repo: https://github.com/OpenSpace/OpenSpace-WebGuiFrontend/pull/175

alexanderbock commented on src/navigation/pathnavigator.cpp:
```diff
@@ -106,6 +106,15 @@ namespace {
+        "JumpToFadeDuration",
+        "JumpTo Fade Duration",
+        "The number of seconds the fading of the rendering should take per default when "
+        "jumping to a scene graph node using the 'openspace.pathnavigation.jumpTo' "
```
Should we mention the script explicitly in here? I think naming the _feature_ instead would be a bit more understandable, since a normal user wouldn't interact with the script necessarily.  Also we might want to use the same value for other jump-like fadings, too

WeirdRubberDuck replied:
Good point. Updated the description. 

Then, maybe we should already apply it to for example jumping to a lat long position on a globe? or a navigation state?

In that case, this property should probably be moved to the navigationhandler instead of the pathnavigator, to make it clear that it can be used for multiple types of transitions. Thoughts?

alexanderbock commented on src/navigation/pathnavigator_lua.inl:
```diff
@@ -307,6 +307,44 @@ namespace {
+                                 std::optional<double> fadeDuration)
+{
+    using namespace openspace;
+    if (!sceneGraphNode(nodeIdentifier)) {
```
```suggestion
    if (SceneGraphNode* n = sceneGraphNode(nodeIdentifier);  !n) {
```

WeirdRubberDuck replied:
Curious: what's the motivation behind this?

alexanderbock replied:
Trying to limit the amount of "work" that is done inside the if statement.  Moving from 1. calling `sceneGraphNode` and 2. testing the result against `nullptr` by making it two separate steps

alexanderbock commented on src/navigation/pathnavigator_lua.inl:
```diff
@@ -307,6 +307,44 @@ namespace {
+    );
+    // No syncing, as this was called from a script that should have been synced already
+    global::scriptEngine->queueScript(
+        script,
```
```suggestion
        std::move(script),
```

WeirdRubberDuck commented on src/navigation/pathnavigator_lua.inl:
```diff
@@ -307,6 +307,44 @@ namespace {
+        global::navigationHandler->pathNavigator().jumpToFadeDuration()
+    );
+
+    std::string onArrivalScript = fmt::format(
```
should now be std::format

WeirdRubberDuck commented:
@alexanderbock  and @ylvaselling I did some small updates. Please give this code another quick look 🙏 

1. Moved the property to NavigationHandler, as more general use cases were expected down the line
1. Created a helper function for fading transitions in the navigation handler, that can be used to fade before and after any transition (or any script, really.  Maybe we should generalize this and e) 
1. Added a jumpTo function for navigation states, using the function above

alexanderbock commented on src/navigation/navigationhandler.cpp:
```diff
@@ -100,6 +109,7 @@ NavigationHandler::NavigationHandler()
     , _disableMouseInputs(DisableMouseInputInfo, false)
     , _disableJoystickInputs(DisableJoystickInputInfo, false)
     , _useKeyFrameInteraction(FrameInfo, false)
+    , _jumpToFadeDuration(JumpToFadeDurationInfo, 1.f, 0.0f, 10.f)
```
```suggestion
    , _jumpToFadeDuration(JumpToFadeDurationInfo, 1.f, 0.f, 10.f)
```

alexanderbock commented on src/navigation/navigationhandler.cpp:
```diff
@@ -178,6 +193,30 @@ void NavigationHandler::setInterpolationTime(float durationInSeconds) {
+    );
+    // No syncing, as this was called from a script that should have been synced already
+    global::scriptEngine->queueScript(
+        script,
```
```suggestion
        std::move(script),
```

alexanderbock commented on src/navigation/pathnavigator_lua.inl:
```diff
@@ -311,8 +311,53 @@ namespace {
+
+    // When copy pasting a printed navigation state from the console, The formatting of
+    // the navigation state dictionary won't be completely correct if using the
+    // dictionary directly , due to the number keys for arrays. We solve this by first
```
```suggestion
    // dictionary directly, due to the number keys for arrays. We solve this by first
```

WeirdRubberDuck commented on src/navigation/pathnavigator_lua.inl:
```diff
@@ -307,6 +307,81 @@ namespace {
+        ));
+    }
+
+    // When copy pasting a printed navigation state from the console, The formatting of
```
```suggestion
    // When copy pasting a printed navigation state from the console, the formatting of
```

WeirdRubberDuck commented on src/navigation/navigationhandler.cpp:
```diff
@@ -178,6 +193,30 @@ void NavigationHandler::setInterpolationTime(float durationInSeconds) {
+{
+    const float duration = fadeDuration.value_or(_jumpToFadeDuration);
+
+    std::string onArrivalScript = std::format(
+        "{} "
+        "openspace.setPropertyValueSingle("
+        "'RenderEngine.BlackoutFactor', 1, {}, 'QuadraticEaseIn'"
+        ")", transitionScript, duration
+    );
+    std::string script = std::format(
```
```suggestion
    const std::string onArrivalScript = std::format(
        "{} "
        "openspace.setPropertyValueSingle("
        "'RenderEngine.BlackoutFactor', 1, {}, 'QuadraticEaseIn'"
        ")", transitionScript, duration
    );
    const std::string script = std::format(
```

WeirdRubberDuck commented on src/navigation/pathnavigator_lua.inl:
```diff
@@ -307,6 +307,81 @@ namespace {
+
+    if (fadeDuration.has_value()) {
+        global::navigationHandler->triggerFadeToTransition(
+            script,
```
```suggestion
            std::move(script),
```

WeirdRubberDuck commented on src/navigation/pathnavigator_lua.inl:
```diff
@@ -307,6 +307,81 @@ namespace {
+
+    if (fadeDuration.has_value()) {
+        global::navigationHandler->triggerFadeToTransition(
+            script,
```
```suggestion
            std::move(script),
```
